waspc/src/Wasp/Analyzer/Parser/Parser.hs

```diff
@@ -0,0 +1,309 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Wasp.Analyzer.Parser.Parser
+  ( -- * AST to CST conversion
+
+    -- | This module takes @["SyntaxNode"]@ produced by 'Wasp.Analyzer.Parser.ConcreteParser'
+    -- and converts it into an abstract syntax tree.
+    parseStatements,
+    parseExpression,
+  )
+where
+
+import Control.Arrow (Arrow (first))
+import Control.Monad.Except (Except, runExcept, throwError)
+import Control.Monad.State.Strict (StateT, evalStateT, gets, modify)
+import Data.Maybe (catMaybes)
+import Wasp.Analyzer.Parser.AST (AST, Expr, ExtImportName, Identifier, Stmt)
+import qualified Wasp.Analyzer.Parser.AST as AST
+import Wasp.Analyzer.Parser.ConcreteParser.CST (SyntaxKind, SyntaxNode (SyntaxNode))
+import qualified Wasp.Analyzer.Parser.ConcreteParser.CST as S
+import Wasp.Analyzer.Parser.Ctx (Ctx (Ctx), WithCtx (WithCtx), ctxFromRgn)
+import Wasp.Analyzer.Parser.ParseError
+import Wasp.Analyzer.Parser.SourcePosition (SourcePosition (SourcePosition))
+import Wasp.Analyzer.Parser.SourceRegion (SourceRegion (SourceRegion))
+import qualified Wasp.Analyzer.Parser.Token as T
+
+data ParseState = ParseState
+  { pstateLine :: !Int,
+    pstateColumn :: !Int,
+    pstateRemainingSource :: String
+  }
+
+pstatePos :: ParseState -> SourcePosition
+pstatePos s = SourcePosition (pstateLine s) (pstateColumn s)
+
+type ParserM a = StateT ParseState (Except ParseError) a
+
+-- | An operation that turns some syntax nodes into a value and outputs the
+-- remaining nodes.
+type Action a = [SyntaxNode] -> ParserM (a, [SyntaxNode])
+
+-- | @parseStatements sourceString syntax@ tries to convert a concrete syntax
+-- tree into an AST.
+parseStatements :: String -> [SyntaxNode] -> Either ParseError AST
```

Comment:
bump

Reply:
The CST represents the entire source file, but it doesn't actually contain the source string (which is needed for getting values for identifiers, strings, numbers, quoters).

Reply:
So CST on its own is not enough to reconstruct the source, is it more like a description of structure of source, which although complete, refers to source at the end.
INteresting, I was actually thinking all the time that CST is a standalone thing, but it turns out it is not. It could be if we stored lexemes in tokens though, right? How did you make this choice?

Reply:
Discussed in a comment chain in `Token.hs`.